File: buildSrc/src/main/java/com/example/AndroidApplicationGradlePlugin.java
//package com.example;
//
//import com.android.build.gradle.AppExtension;
//import com.android.build.gradle.internal.dsl.BuildType;
//
//import org.gradle.api.Project;
//
//public class AndroidApplicationGradlePlugin extends AndroidProjectGradlePlugin {
//    @Override
//    void applyPlugins(Project project) {
//        project.getPluginManager().apply("com.android.application");
//        super.applyPlugins(project);
//
//        System.out.println(project.getProperties());
//    }
//
//    @Override
//    void configurePlugins(Project project) {
//        super.configurePlugins(project);
//        AppExtension androidExtension = project.getExtensions().getByType(AppExtension.class);
//        androidExtension.buildTypes(buildTypes -> {
//            BuildType release = buildTypes.findByName("release");
//            release.setMinifyEnabled(false);
//            release.proguardFiles(
//                    androidExtension.getDefaultProguardFile("proguard-android-optimize.txt"),
//                    "proguard-rules.pro"
//            );
//        });
//    }
//
//    @Override
//    void applyDependencies(Project project) {
//        super.applyDependencies(project);
//    }
//}

File: buildSrc/src/main/kotlin/com/example/AndroidApplicationGradlePlugin.kt
package com.example

import com.android.build.gradle.AppExtension
import org.gradle.api.Project

class AndroidApplicationGradlePlugin : AndroidProjectGradlePlugin() {
    override fun applyPlugins(project: Project) {
        project.pluginManager.apply("com.android.application")
        super.applyPlugins(project)
    }

    override fun configurePlugins(project: Project) {
        super.configurePlugins(project)
        project.extensions.getByType(AppExtension::class.java).apply {
            defaultConfig.apply {
                applicationId = "com.example.modularization"
            }

            buildTypes.apply {
                findByName("release")?.apply {
                    isMinifyEnabled = false
                    proguardFiles(
                        getDefaultProguardFile("proguard-android-optimize.txt"),
                        "proguard-rules.pro"
                    )
                }
            }
        }
    }

    override fun applyDependencies(project: Project) {
        super.applyDependencies(project)
    }
}